# BÀI TẬP BỘ 1/5 — TITANIC
## Data Profiling → Thống kê mô tả đơn biến
**Nguồn:** kaggle.com/c/titanic (891 dòng)

Đi từ Chương 0 (Data Profiling) sang Chương 1 (Phân tích đơn biến) trên cùng 1 dataset.

## Setup

In [10]:
import pandas as pd
df.head(10)


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True
5,0,3,male,NaN,0,0,8.4583,Q,Third,man,True,NaN,Queenstown,no,True
6,0,1,male,54.0,0,0,51.8625,S,First,man,True,E,Southampton,no,True
7,0,3,male,2.0,3,1,21.0750,S,Third,child,False,NaN,Southampton,no,False
8,1,3,female,27.0,0,2,11.1333,S,Third,woman,False,NaN,Southampton,yes,False
9,1,2,female,14.0,1,0,30.0708,C,Second,child,False,NaN,Cherbourg,yes,False


---
# PHẦN A — DATA PROFILING
## Câu 1 — Data size, column names, data types

In [21]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    str     
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    str     
 8   class        891 non-null    category
 9   who          891 non-null    str     
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    str     
 13  alive        891 non-null    str     
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), str(5)
memory usage: 80.7 KB


## Câu 2 — Missing values (count + %), cột nào thiếu nhiều nhất?

In [29]:
df.isnull().sum().sort_values(ascending= False)

deck           688
age            177
embarked         2
embark_town      2
sex              0
pclass           0
survived         0
fare             0
parch            0
sibsp            0
class            0
adult_male       0
who              0
alive            0
alone            0
dtype: int64

## Câu 3 — Duplicate rows

In [ ]:
df.duplicated().sum()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


**Lưu ý:** 107 dòng bị trùng lặp hoàn toàn. Bộ dữ liệu này không có cột ID/tên hành khách nên "trùng" ở đây là trùng toàn bộ giá trị các cột hiện có (ví dụ 2 nam giới cùng 22 tuổi, cùng hạng vé, cùng giá vé...), không chắc là trùng *cùng một người* trong thực tế.

## Câu 4 — Invalid values: age có giá trị âm hoặc > 100 không?

In [20]:
df[(df['age'] < 0) | (df['age'] > 100)]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone


## Câu 5 — Phân loại biến: survived, pclass, sex, age, fare, embarked thuộc loại nào?

- `survived`: định danh nhị phân (0/1)
- `pclass`: định danh có thứ tự (1, 2, 3 — hạng vé)
- `sex`: định danh (nominal)
- `age`: định lượng liên tục
- `fare`: định lượng liên tục
- `embarked`: định danh (nominal)

---
# PHẦN B — THỐNG KÊ MÔ TẢ ĐƠN BIẾN
## B.1. Biến định lượng: age
### Câu 6 — Mean, Median, Std, Skewness, Kurtosis

### Câu 7 — Histogram + Boxplot cho age

## B.2. Biến định lượng: fare
### Câu 8 — Mean, Median, Std, Skewness, Kurtosis, nhận xét độ lệch

**Nhận xét:** `fare` lệch phải rất mạnh (skew ≈ 4.79, kurtosis ≈ 33.4) — trung bình (~32.2) cao hơn hẳn median (~14.45) vì một số ít vé rất đắt kéo trung bình lên. Nên dùng **median** để mô tả fare thay vì mean.

### Câu 9 — Phát hiện outlier bằng IQR cho fare

## B.3. Biến định danh: survived
### Câu 10 — Tần suất, tỷ lệ % sống sót

## B.4. Biến định danh: sex, pclass, embarked
### Câu 11 — Tần suất + tỷ lệ % cho từng biến, Mode của mỗi biến

### Câu 12 — Vẽ bar chart cho pclass

---
# PHẦN C — TỔNG HỢP
## Câu 13 — describe() cho age, fare
## Câu 14 — Viết 4-5 câu insight tổng hợp

**Insight tổng hợp:** Dữ liệu gồm 891 dòng, 15 cột, thiếu nhiều nhất ở `deck` (~77%, gần như bỏ được cột này) và `age` (~20%, cần cân nhắc điền khuyết thay vì loại bỏ). Tỷ lệ sống sót chỉ **38.4%**, đa số hành khách (61.6%) không qua khỏi. `age` có phân phối gần đối xứng (skew ≈ 0.39, tuổi trung bình ~29.7), trong khi `fare` lệch phải rất mạnh (skew ≈ 4.79) với khoảng **13%** giá trị là outlier — nên ưu tiên median khi mô tả fare. Hành khách trong dữ liệu chủ yếu là nam (64.8%), đi vé hạng 3 (55.1%) và lên tàu tại Southampton (72.4%), cho thấy các nhóm này chiếm áp đảo và cần lưu ý khi so sánh tỷ lệ sống sót giữa các nhóm ở phần phân tích tiếp theo.